src/resolvelib/resolvers.py

```diff
@@ -173,6 +173,33 @@ def _add_to_criteria(self, criteria, requirement, parent):
             raise RequirementsConflicted(criterion)
         criteria[identifier] = criterion
 
+    def _remove_information_from_criteria(self, criteria, parents):
+        """
+        Removes information from a set of parents from a criteria. Concretely, removes
+        from each criterion's `information` field all values that have one of `parents`
+        as provider of the requirement.
+
+        :param criteria: The criteria to update.
+        :param parents: The set of identifiers for which to remove information from all
+            criteria.
+        """
+        if not parents:
+            return
+        for key, criterion in criteria.items():
+            criteria[key] = Criterion(
+                criterion.candidates,
+                # TODO: is empty information allowed?
```

Comment:
Empty information is technically _allowed_ but likely unexpected, since it records where a criterion came from, and if a criterion comes from nowhere, should it not be removed?

Reply:
My reasoning was that we're dropping the constraint provided by a parent because we're dropping the pin on that parent. But if the new pin on the parent happens to have the same constraint (the common case), we might want to keep `incompatibilities`, in order to prevent backtracking over the same thing again.
But as I type this I realize that I'm really not yet confident in my understanding of how incompatibilities are gathered. So if you are confident that we can safely drop the criterion when no information remains I'll apply that change. Otherwise I'll have to sink my teeth in the incompatibilities some more.

Reply:
If the intent is to keep `incompatibilities` in the graph, I think it’s OK to keep a criterion with empty information (it’s basically an orphan node). We should describe this in the release note (by adding a release note fragment file in `news`) so provider/reporter implementers are aware of the possibility that `information` _could_ be empty (and what that means), it’s likely no-one would be particular troubled by this since an orphan node can simply be ignored in the final graph anyway.

Reply:
All right, I'll do that. I think `pip` will need a minor change, but there seems to already be a breaking change in master compared to the last release anyway so I guess that is acceptable.

Reply:
Does the pin need to be excluded from the final result whose criterion has empty information? Since `_is_current_pin_satisfying()` will always return True if there has been one pinned before.

Reply:
I think the orphan nodes will be excluded automatically by `build_result()`

Reply:
I've made an attempt at the news fragment but I'm not entirely sure about your change entry conventions.